pageserver/src/layered_repository.rs

```diff
@@ -431,19 +396,29 @@ impl Repository {
         Ok(())
     }
 
-    /// Updates timeline based on the `TimelineSyncStatusUpdate`, received from the remote storage synchronization.
-    /// See [`crate::remote_storage`] for more details about the synchronization.
-    pub fn attach_timeline(&self, timeline_id: ZTimelineId) -> Result<()> {
-        debug!("attach timeline_id: {}", timeline_id,);
-        match self.timelines.lock().unwrap().entry(timeline_id) {
-            Entry::Occupied(_) => bail!("We completed a download for a timeline that already exists in repository. This is a bug."),
-            Entry::Vacant(entry) => {
-                // we need to get metadata of a timeline, another option is to pass it along with Downloaded status
-                let metadata = load_metadata(self.conf, timeline_id, self.tenant_id).context("failed to load local metadata")?;
-                // finally we make newly downloaded timeline visible to repository
-                entry.insert(LayeredTimelineEntry::Unloaded { id: timeline_id, metadata })
-            },
-        };
+    pub fn init_attach_timelines(
+        &self,
+        timelines: Vec<(ZTimelineId, TimelineMetadata)>,
+    ) -> anyhow::Result<()> {
+        let sorted_timelines = tree_sort_timelines(timelines)?;
```

Comment:
It looks from the code, that it is ok with 1 element "sorting", but sure, adjusted the code anyway.

Reply:
It will fail if passed timeline has an ancestor but ancestor is missing from the list

This will panic:

```rust
let metadata = TimelineMetadata::new(
    Lsn(0),
    None,
    Some(ZTimelineId::generate()),
    Lsn(0),
    Lsn(0),
    Lsn(0)
);
tree_sort_timelines(vec![(ZTimelineId::generate(), metadata)]).unwrap();
```